### PR TITLE
feat: restructure caches and their relationship with syncFunction (playout only)

### PR DESCRIPTION
**Pull request by Julusian**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

feature/refactoring

subset of https://github.com/nrkno/tv-automation-server-core/pull/354

* **What is the current behavior?** (You can also link to an open issue here)

Cache and ingest shares a cache type. This is not optimal, as it means that ingest is having to load in the data for the whole playlist to do updates on just one. It also means that playout is loading in data it will never need.
Additionally they are forced to share a lock(syncFunction) because of this. This is bad as a slow ingest operation can block a take from taking effect.

* **What is the new behavior (if this is a feature change)?**

No major functional changes, this is laying the groundwork for improving this.
The playout logic has been moved to a new cache type, and new syncfunctions that are designed for that cache.
Additionally, a new syncFunction layer has been added for studio operations. This encompasses all the playlist syncFunctions, but also activating/deactivating a playout, and some studio baseline operations. This is to reduce risk of error if these operations happen to collide.

The new lockFunctions are designed to be the sole way to obtain the appropriate cache. They are saved when the function returns. They simply need the playlistId and they will load the data needed for playout. There is a preInit step where some playlist validation can be done to filter out clearly invalid calls, to avoid loading the full set of data. (eg doing a take when the playlist is not active)

* **Other information**:

The linter rule `no-floating-promises` has been enabled due to the increased use of promises in this code. It flagged a few relevant issues.

**Status**

- [ ] Code documentation for the relevant parts in the code have been added/updated by the PR author
- [ ] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

File: meteor/server/api/playout/cache.ts [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+import { ActivationCache, getActivationCache } from '../../cache/ActivationCache'
+import { DbCacheReadCollection, DbCacheWriteCollection } from '../../cache/CacheCollection'
+import { DbCacheReadObject, DbCacheWriteObject } from '../../cache/CacheObject'
+import { CacheBase, CacheForRundownPlaylist } from '../../cache/DatabaseCaches'
+import { profiler } from '../profiler'
+import { removeRundownPlaylistFromDb } from '../rundownPlaylist'
+import { CacheForStudioBase } from '../studio/cache'
+import { getRundownsSegmentsAndPartsFromCache } from './lib'
+
+export abstract class CacheForPlayoutPreInit extends CacheBase<CacheForPlayout> {
```
Comment (nytamin) [inline meteor/server/api/playout/cache.ts:27]: doc: Could you write some documentation on what the intended use for the various caches are?

File: meteor/server/api/playout/syncFunction.ts [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+/**
+ * Lock the playlist for performing a playout operation and load a cache of data.
+ * Note: This also locks the studio, if that is already locked then use 'rundownPlaylistPlayoutFromStudioLockFunction' instead
+ * @param context Meteor call context, to authenticate the request. Pass null if already authenticated
+ * @param contextStr Contextual information for the call to this function. to aid debugging
+ * @param rundownPlaylistId Id of the playlist to lock
+ * @param fcn Function to run while holding the lock
+ */
+export function playoutWithCacheLockFunction<T>(
```
Comment (nytamin) [inline meteor/server/api/playout/syncFunction.ts:39]: Just a minor suggestion: Could these functions be renamed into something more imperative? I don't know, perhaps `runWithPlayoutCacheLock` or something?

**Review by nytamin — COMMENTED** [review]
Looks really good!

**Comment by Julusian** [comment]
There are some remaining linter errors revealed here:
```
/home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/RundownList.tsx:238:5
ERROR: 238:5  no-floating-promises  Promises must be handled appropriately

/home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/RundownList/RundownListItem.tsx:234:6
ERROR: 234:6  no-floating-promises  Promises must be handled appropriately

/home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/RundownList/RundownPlaylistUi.tsx:145:6
ERROR: 145:6  no-floating-promises  Promises must be handled appropriately
ERROR: 155:6  no-floating-promises  Promises must be handled appropriately
ERROR: 165:5  no-floating-promises  Promises must be handled appropriately

/home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/Settings.tsx:185:6
ERROR: 185:6  no-floating-promises  Promises must be handled appropriately
ERROR: 206:6  no-floating-promises  Promises must be handled appropriately
ERROR: 227:6  no-floating-promises  Promises must be handled appropriately

/home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/i18n.ts:101:4
ERROR: 101:4  no-floating-promises  Promises must be handled appropriately
```
@jstarpl I think they are all places where we do a Meteor.call, and havent put a catch on the promise. I can go though and add a basic one that logs to the console unless you want to do something better?

File: meteor/server/api/playout/cache.ts (hunk 1 above)
Comment (Julusian) [reply meteor/server/api/playout/cache.ts:31]: done

**Comment by jstarpl** [comment]
> 
> 
> There are some remaining linter errors revealed here:
> 
> ```
> /home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/RundownList.tsx:238:5
> ERROR: 238:5  no-floating-promises  Promises must be handled appropriately
> 
> /home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/RundownList/RundownListItem.tsx:234:6
> ERROR: 234:6  no-floating-promises  Promises must be handled appropriately
> 
> /home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/RundownList/RundownPlaylistUi.tsx:145:6
> ERROR: 145:6  no-floating-promises  Promises must be handled appropriately
> ERROR: 155:6  no-floating-promises  Promises must be handled appropriately
> ERROR: 165:5  no-floating-promises  Promises must be handled appropriately
> 
> /home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/Settings.tsx:185:6
> ERROR: 185:6  no-floating-promises  Promises must be handled appropriately
> ERROR: 206:6  no-floating-promises  Promises must be handled appropriately
> ERROR: 227:6  no-floating-promises  Promises must be handled appropriately
> 
> /home/julus/Projects/sofietv/server-core/mono-repo/meteor/client/ui/i18n.ts:101:4
> ERROR: 101:4  no-floating-promises  Promises must be handled appropriately
> ```
> 
> @jstarpl I think they are all places where we do a Meteor.call, and havent put a catch on the promise. I can go though and add a basic one that logs to the console unless you want to do something better?

Actually, all of these probably need something like the `doUserAction` wrapper, just one that's more generic, since it would make sense to show a proper notification message if those actions fail.

**Review by sparkpunkd — APPROVED** [review]
This is the beginning of a journey and this separation of cache type ... and hence when locking is required and for what ... is a step towards Sofie working more smoothly. So, as it stands, I approve this PR.

However, as a more general comment, solving the problem of operations being delayed by adding more kinds of locks and more specialized caches - which in turn adds more code, maintenance overhead and still blocks operations. This PR touches 73 files! This is not the correct arhcitectural approach long term.

**Comment by jstarpl** [comment]
@Julusian should I write a UI wrapper to handle those promises and show a notification if they fail?

**Comment by Julusian** [comment]
@jstarpl I see there already is one, so I shall switch those userAction ones over.
Looking again, a few of them are calls like `MeteorCall.blueprint.removeBlueprint(blueprint._id)`. Further up we are doing `MeteorCall.studio.insertStudio().catch(console.error)`, so I shall add the same catch for now. Feel free to make something better, but I don't expect it to be worth the effort to do for now

**Comment by Julusian** [comment]
Merged as part of #463

**Pull request closed without merging**